command/select.go

```diff
@@ -0,0 +1,212 @@
+package command
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/urfave/cli/v2"
+
+	errorpkg "github.com/peak/s5cmd/error"
+	"github.com/peak/s5cmd/log/stat"
+	"github.com/peak/s5cmd/parallel"
+	"github.com/peak/s5cmd/storage"
+	"github.com/peak/s5cmd/storage/url"
+)
+
+var selectHelpTemplate = `Name:
+	{{.HelpName}} - {{.Usage}}
+
+Usage:
+	{{.HelpName}} [options] argument
+
+Options:
+	{{range .VisibleFlags}}{{.}}
+	{{end}}
+Examples:
+	01. Search for all JSON objects with the foo property set to 'bar' and spit them into stdout
+		 > s5cmd {{.HelpName}} --compression gzip --query "SELECT * FROM S3Object s WHERE s.foo='bar'" s3://bucket/*
+`
+
+var selectCommandFlags = []cli.Flag{
+	&cli.StringFlag{
+		Name:    "query",
+		Aliases: []string{"e"},
+		Usage:   "SQL expression to use to select from the objects",
+	},
+	&cli.StringFlag{
+		Name:  "compression",
+		Usage: "input compression format",
+		Value: "NONE",
+	},
+	&cli.StringFlag{
+		Name:  "format",
+		Usage: "input data format (only JSON supported for the moment)",
+		Value: "JSON",
+	},
+}
+
+var selectCommand = &cli.Command{
+	Name:               "select",
+	HelpName:           "select",
+	Usage:              "run SQL queries on objects",
+	Flags:              selectCommandFlags,
+	CustomHelpTemplate: selectHelpTemplate,
+	Before: func(c *cli.Context) error {
+		err := validateSelectCommand(c)
+		if err != nil {
+			printError(givenCommand(c), c.Command.Name, err)
+		}
+		return err
+	},
+	Action: func(c *cli.Context) (err error) {
+		defer stat.Collect(c.Command.FullName(), &err)()
+
+		return Select{
+			src:         c.Args().Get(0),
+			op:          c.Command.Name,
+			fullCommand: givenCommand(c),
+			// flags
+			query:           c.String("query"),
+			compressionType: c.String("compression"),
+
+			storageOpts: NewStorageOpts(c),
+		}.Run(c.Context)
+	},
+}
+
+// Select holds select operation flags and states.
+type Select struct {
+	src         string
+	op          string
+	fullCommand string
+
+	query           string
+	compressionType string
+
+	// s3 options
+	storageOpts storage.Options
+}
+
+// Run starts copying given source objects to destination.
+func (s Select) Run(ctx context.Context) error {
+	srcurl, err := url.New(s.src)
+	if err != nil {
+		printError(s.fullCommand, s.op, err)
+		return err
+	}
+
+	client, err := storage.NewRemoteClient(ctx, srcurl, s.storageOpts)
+	if err != nil {
+		printError(s.fullCommand, s.op, err)
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	objch, err := expandSource(ctx, client, false, srcurl)
+	if err != nil {
+		printError(s.fullCommand, s.op, err)
+		return err
+	}
+
+	var merror error
+
+	waiter := parallel.NewWaiter()
+	errDoneCh := make(chan bool)
+	writeDoneCh := make(chan bool)
```

Comment:
If I'm not missing something, this channel is not used. It is closed by goroutine but we don't check whether it is closed. Shouldn't we need to wait in parent just like we did in`errDoneCh`?

```go
waiter.Wait()
<-errDoneCh
<-writeDoneCh

return merror
```

Reply:
Yes, I think you're right.